README.md
# Table Of Contents
- Introduction
- Features
- Documentation
- Prerequisites
- Install
- Contributions
- License

## Introduction

Find A Bike Dublin is a web application designed to allow users to find the best Dublin Bikes Station in their vicinity to either rent out a bike or leave a bike back.  To launch an instance, just clone it, navigate to the DublinBIkesApp folder and type flask run.  Due to our repsonsive design, the application is easily viewable on all device from laptops to phones and everything in between.

## Features

- __Real time bike info:__ Find information both on how many bikes are available at each bikestand and how many bike stands are also available. 

- __Map showing bike stand locations:__ Not only does the app provide information on the number of bikes and bikestands available, through google maps integration, it also displays their location on a map, allowing the user to quickly locate the nearest bike station suitable for their needs.

- __Expertly crafted GUI:__ The GUI has been crafted to be both easy to understand and intuitive to use.  It's design has been kept as minimal as possible so that the user can access information about the various bike stations dotted around Dublin as quickly as possible.

## Documentation

- Comprehensive documentation on the use of the app is available to each user via a separate webpage present in the app

- Other documentation can also be found in the quick start guide and FAQ also available on git hub.

## Prerequisites

### Flask

Find A Bike Dublin requires flask 1.0.2 or above. It also makes use of several flask expansions which are listed below:

- flask-sqlalchemy 2.3.2
- flask-migrate 2.4.0
- flask-wtf 0.14.2

### SQLAlchemy

Find A Bike Dublin also requires sqlalchemy 1.3.0 for sql operations.

### Anaconda Virtual Environment

Find A Bike Dublin makes use of Conda virtual environments, the version used is 4.6.2.

## Installation

1. Clone the git hub repository with the following url:
        $ git clone https://github.com/Moss89/DublinBikesProject.git
2. Navigate to the DublinBikesApp folder:
        $ cd /DublinBikesProject/DublinBikesApp
3. Pip install the requirements.txt file:
        $ pip install -r requirements.txt
4. Run the flask app:
        $ flask run

## Contributions

- John Hackett
- Tomas Murphy
- John McLoughlin

## License

The license is GPL3 for all parts specific to Find A Bike Dublin, this includes:

- The core files
- The documentaion


changeLog.md
commit a6072b5be57936cf6af03f5cbd2286668bdf81de
Author: jmcl <john.mc-loughlin.3@ucdconnect.ie>
Date:   Fri Apr 19 11:50:55 2019 +0100

    added machine learning notebook with csv needed to build models

commit 4e99f85a4edfe74c41a2c8a279c514c8d718ab9a
Author: jmcl <john.mc-loughlin.3@ucdconnect.ie>
Date:   Fri Apr 19 11:41:34 2019 +0100

    tried to make changelog more readable

commit 549f8d0ed9c3864872a4c6ac3cd77b18fa809d08
Author: jmcl <john.mc-loughlin.3@ucdconnect.ie>
Date:   Fri Apr 19 11:38:59 2019 +0100

    update readme and changelog

commit 28aba1274d80f63d2275c2254060f7a1eff3ef93
Author: jmcl <john.mc-loughlin.3@ucdconnect.ie>
Date:   Fri Apr 19 11:37:07 2019 +0100

    README and Changelog updated

commit 90219db34e3638f30b1366d47565ae53c982fe10
Author: jmcl <john.mc-loughlin.3@ucdconnect.ie>
Date:   Fri Apr 19 11:32:01 2019 +0100

    Removed more unecessary files

commit a9c283d7fe24c1709b239eee40897ad6de7e16c8
Author: jmcl <john.mc-loughlin.3@ucdconnect.ie>
Date:   Fri Apr 19 11:28:28 2019 +0100

    removed DS file

commit 6e43496529795b5dafdf157825da1bd21df13a09
Author: jmcl <john.mc-loughlin.3@ucdconnect.ie>
Date:   Fri Apr 19 11:27:58 2019 +0100

    Removed flask app practice file

commit 80808283c2bc85bd7880c33dda66c8b51402c137
Author: jmcl <john.mc-loughlin.3@ucdconnect.ie>
Date:   Fri Apr 19 11:27:10 2019 +0100

    Removed website design

commit 18851c667eb2c0afba12f117f24c753e1dbb4e23
Author: jmcl <john.mc-loughlin.3@ucdconnect.ie>
Date:   Fri Apr 19 11:26:17 2019 +0100

    Removed .idea directory

commit 93ae4fa4ca4db732ab6188bdda8bc6a972033b3f
Author: jmcl <john.mc-loughlin.3@ucdconnect.ie>
Date:   Fri Apr 19 11:25:31 2019 +0100

    Removed practice directory

commit 6a8bc29b96e4c02a6b511678c37927381211c91e
Merge: 2772084 16c0f5e
Author: jmcl <john.mc-loughlin.3@ucdconnect.ie>
Date:   Fri Apr 19 11:17:52 2019 +0100

    Merge to master

commit 277208494838eec632b70cb53f6c1feb71f5d537
Author: Tom <Tomas.Murphy2@gmail.com>
Date:   Fri Apr 19 10:10:13 2019 +0100

    Clean up of code.
    
    Signed-off-by: Tom <Tomas.Murphy2@gmail.com>

commit 3113b30d032d4a712064a90ba2988b0affbb39dc
Author: Tom <Tomas.Murphy2@gmail.com>
Date:   Fri Apr 19 10:05:16 2019 +0100

    Chart width changed to 100%.
    
    Signed-off-by: Tom <Tomas.Murphy2@gmail.com>

commit 40bf4decd4a2c8182102d7c57237e1a477c992b1
Author: Tom <Tomas.Murphy2@gmail.com>
Date:   Fri Apr 19 09:35:47 2019 +0100

    Fixed issue with website scaling on EC2 instance
    
    Signed-off-by: Tom <Tomas.Murphy2@gmail.com>

commit dd3161495c4191f8d782ea055207486fcca50859
Author: john <john.hackett1@ucdconnect.ie>
Date:   Fri Apr 19 09:15:22 2019 +0100

    Animation change, colour alteration on graph

commit be05b30697b9340c4dbc85927446fb40ae2907be
Author: jmcl <john.mc-loughlin.3@ucdconnect.ie>
Date:   Fri Apr 19 08:32:47 2019 +0100

    Unit test file added
    get bike info tidied up
    predictor doesn't run for past now
    requirements.txt updated
    Removed unecessary lines
    
    Signed-off-by: jmcl <john.mc-loughlin.3@ucdconnect.ie>

commit 97c185ba6bd4bc6fe075a31b58c2ab11d3f3e204
Author: jmcl <john.mc-loughlin.3@ucdconnect.ie>
Date:   Thu Apr 18 21:26:15 2019 +0100

    5 day limit imposed
    
    Signed-off-by: jmcl <john.mc-loughlin.3@ucdconnect.ie>

commit 51db48000f5c8d1e7e266c255adf88274c7c1e05
Author: jmcl <john.mc-loughlin.3@ucdconnect.ie>
Date:   Thu Apr 18 19:38:57 2019 +0100

    Fixed 3 hr issue with prediction
    
    Signed-off-by: jmcl <john.mc-loughlin.3@ucdconnect.ie>

commit 09448feaefeefa3a27c3728b550ea9b95a108f1a
Author: Tom <Tomas.Murphy2@gmail.com>
Date:   Thu Apr 18 18:09:50 2019 +0100

    Predeployment test
    
    Signed-off-by: Tom <Tomas.Murphy2@gmail.com>

commit 4a3c4dbc5f624fe0ded41ea01849168aa9fd15de
Author: john <john.hackett1@ucdconnect.ie>
Date:   Wed Apr 17 21:25:36 2019 +0100

    InfoWindow updating properly

commit 4d97e67d2003842ac72d9c06b4852b8b4af7d9df
Author: john <john.hackett1@ucdconnect.ie>
Date:   Wed Apr 17 20:54:55 2019 +0100

    Chart now shows new graph on predictions
    
    When a user predicts a day,The chart updates to show that specific days most recent bike history

commit d3a2b6efc10e43ee94ea01d80f719a43e53fb007
Author: john <john.hackett1@ucdconnect.ie>
Date:   Tue Apr 16 01:26:14 2019 +0100

    Chart improvement

commit b853e249bb5e6a47f3ba3fbd24a1056311c9eba2
Author: john <john.hackett1@ucdconnect.ie>
Date:   Sat Apr 13 12:52:15 2019 +0100

    Markers now updating with predictions

commit 2550ad25a0567c191020a5e9646003213d49ae9a
Author: jmcl <john.mc-loughlin.3@ucdconnect.ie>
Date:   Fri Apr 12 11:31:04 2019 +0100

    Changed format for output from prediction function
    
    Signed-off-by: jmcl <john.mc-loughlin.3@ucdconnect.ie>

commit 96f20a4d9d3399b2e6190ef9faed6eb543a6aa60
Author: jmcl <john.mc-loughlin.3@ucdconnect.ie>
Date:   Thu Apr 11 10:45:37 2019 +0100

    MAchine learning model now get datetime and weather info
    Just needs to output results to map
    
    Signed-off-by: jmcl <john.mc-loughlin.3@ucdconnect.ie>

commit 3a4932dd8a05da3ad77fff54c5c83e961cfce805
Author: john <john.hackett1@ucdconnect.ie>
Date:   Tue Apr 9 20:22:50 2019 +0100

    Change to how date time is processed
    
    Date time turned to epoch time for ml step

commit 2bf6e7587dee932c7b2f8533e1b056634f738b3b
Author: jmcl001 <43282116+jmcl001@users.noreply.github.com>
Date:   Tue Apr 9 09:09:27 2019 +0100

    ReadMe updated, let me know what you think

commit 1e646285fbab04e258b8349671c1124df0006108
Author: john <john.hackett1@ucdconnect.ie>
Date:   Mon Apr 8 20:35:02 2019 +0100

    DublinBikes_js changed
    
    Should've been added in previous commit

commit e89b3b234be8cf9f04111718eb648c449aedbd79
Author: john <john.hackett1@ucdconnect.ie>
Date:   Mon Apr 8 20:32:19 2019 +0100

    Integration completed for Tom and JohnH work
    
    Finished website integrated into work. Some features still need polish.


DublinBikesApp/helpers.py
from flask import request
import datetime
from requests import get
from json import loads

"""
Functions for the routes go here
"""


def get_static_data(table):

    results = {"number": [], "lat": [], "lng": [], "address": [], "bikestands": []}
    for i in range(len(table)):
        results["number"].append(table[i].number)
        results["lat"].append(table[i].lat)
        results["lng"].append(table[i].lng)
        results["address"].append(table[i].address)
        results["bikestands"].append(table[i].bikestands)
    return results


def get_date_time():
    date = request.json['date']
    time = request.json['time']
    date_time = datetime.datetime(int(date[0:4]), int(date[5:7]), int(date[8:10]), int(time[0:2]), int(time[3:5]))
    weekday = date_time.weekday()
    return date_time, weekday


def get_lat_lng():
    # Rounded to 6 decimal points to match with database lat and lng data entry.
    lat = round(request.json['lat'], 7)
    lng = round(request.json['lng'], 7)
    return lat, lng


def getWeather(t):

    # Scrape weather API
    r = get(
        "https://api.openweathermap.org/data/2.5/forecast/?q=dublin&APPID=718d1e3d695907c31b9a4b710e8348f5&units=metric&cnt=200&mode=json").text
    # Convert scrape to python dict
    r = loads(r)

    # Iterate over weather info and get relevant weather
    for i in range(1, len(r["list"])):
        if r["list"][i - 1]["dt"] < t < r["list"][i]["dt"]:
            return r["list"][i]

    # Used if search up to 3 hours from current
    return r["list"][0]


def formatWeatherInfo(weather, date_time, day):
    # Dictionaries for converting weather strings to numeric representation
    descriptionDict = {'broken clouds': 0,
                       'scattered clouds': 1,
                       'few clouds': 2,
                       'clear sky': 3,
                       'mist': 4,
                       'fog': 5,
                       'light rain': 6,
                       'moderate rain': 7,
                       'heavy intensity rain': 8,
                       'very heavy rain': 9,
                       'snow': 10,
                       'light intensity shower rain': 11,
                       'shower rain': 12,
                       'light intensity drizzle': 13,
                       'shower sleet': 14,
                       'light intensity drizzle rain': 15,
                       'drizzle': 16,
                       'overcast clouds': 17}

    overviewDict = {'Rain': 0,
                    'Clouds': 1,
                    'Fog': 2,
                    'Snow': 3,
                    'Clear': 4,
                    'Mist': 5,
                    'Drizzle': 6}

    # Formatting info for passing to model
    overview = weather["weather"][0]["main"]
    description = weather["weather"][0]["description"]
    temp = weather["main"]["temp"]
    windSpeed = weather["wind"]["speed"]
    clouds = weather["clouds"]["all"]
    info = [date_time, day, overview, description, temp, windSpeed, clouds]

    # In case weather unknown at the time encountered use this
    if overview not in overviewDict.keys():
        info[2] = 20
    else:
        info[2] = overviewDict[info[2]]
    if description not in descriptionDict.keys():
        info[3] = 20
    else:
        info[3] = descriptionDict[info[3]]

    return info


def formatDateTime(date_time):
    day = date_time[1]
    date_time = int(date_time[0].strftime('%s'))

    return date_time, day